src/build_network.py

```diff
@@ -0,0 +1,97 @@
+
+    Note:
+    This function is based on code from Le et al., and used in accordance with their license:
+    https://github.com/greenelab/iscb-diversity/blob/master/02.process-pubmed.ipynb
+    '''
+    # I have no idea why there isn't a better way to get rid of double extensions
+    base_name = os.path.splitext(os.path.splitext(file_path)[0])[0]
+    pickle_path = base_name + '.pkl'
+
+    # For speed, load the pickled version if it's available
+    if os.path.exists(pickle_path):
+        with open(pickle_path, 'rb') as in_file:
+            article_df = pickle.load(in_file)
+    else:
+        articles = []
+        # generator of XML PubmedArticle elements
+        article_elems = iter_extract_elems(file_path, tag='PubmedArticle')
+
+        for elem in article_elems:
+            # Example efetch XML for <PubmedArticle> at https://github.com/dhimmel/pubmedpy/blob/f554a06e13e24d661dc5ff93ad07179fb3d7f0af/pubmedpy/data/efetch.xml
+            articles.append(extract_all(elem))
+
+        article_df = pd.DataFrame(articles)
+
+        with open(pickle_path, 'wb') as out_file:
+            pickle.dump(article_df, out_file)
+
+    return article_df
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--data_dir',
+                         help='The directory containing coci citations',
+                         default='data/coci')
+    parser.add_argument('--metadata_dir',
+                        help='File from download_article_metadata containing info on the '
+                             'articles to add to the network',
+                        default='data/pubmed/efetch')
+    parser.add_argument('--out_dir',
+                        help='The location to save the resulting netoworks to',
+                        default='data/networks')
+    args = parser.parse_args()
+
+    print('Initializing graphs...')
+    metadata_files = glob.glob(f'{args.metadata_dir}/*.xz')
+    heading_to_graph = {}
+    heading_to_dois = {}
+    headings = []
+    for metadata_path in metadata_files:
+        heading = os.path.basename(metadata_path)
+        heading = heading.split('.')[0]
+        headings.append(heading)
+        heading_to_graph[heading] = nx.DiGraph()
+
+        article_df = parse_metadata(metadata_path)
+        dois = set(article_df['doi'])
+        heading_to_dois[heading] = dois
+
+    print('Building graphs...')
+    # For reasons that are unclear to me, this is faster than either inserting all
+    # edges at once or constructing the graph in one shot
+    for file_path in tqdm(glob.glob(f'{args.data_dir}/*')):
+        citation_list = pd.read_csv(file_path)
+        for heading in headings:
+            heading_dois = heading_to_dois[heading]
+            for citing, cited in zip(citation_list['citing'], citation_list['cited']):
+                if citing in heading_dois and cited in heading_dois:
+                    heading_to_graph[heading].add_edge(citing, cited)
```

Comment:
Does this mean you're only counting citations within headings (i.e. from a comp bio paper to another comp bio paper, but not from a comp bio paper to an ML paper)? I wonder how many citations aren't under the same heading, or if there's a way to count that.